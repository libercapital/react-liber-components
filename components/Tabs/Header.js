'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Header = function Header(_ref) {
  var children = _ref.children,
      labels = _ref.labels,
      slider = _ref.slider,
      header = _ref.header,
      asList = _ref.asList;
  return _react2.default.createElement(
    _react.Fragment,
    null,
    asList ? children(labels, slider, header) : children(labels)
  );
};

Header.propTypes = {
  children: _propTypes2.default.func,
  labels: _propTypes2.default.oneOfType([_propTypes2.default.any]),
  slider: _propTypes2.default.oneOfType([_propTypes2.default.any]),
  header: _propTypes2.default.oneOfType([_propTypes2.default.any]),
  asList: _propTypes2.default.bool
};

Header.defaultProps = {
  children: function children() {},
  labels: null,
  slider: null,
  header: null,
  asList: false
};

exports.default = Header;