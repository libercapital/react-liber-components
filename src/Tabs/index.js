export {
  Tabs,
  Tab,
  TabHeader,
  SimpleTabs,
  SimpleTab,
  DropDownTab,
  TabContainer,
  TabContent,
  TabContentDisplay,
} from './tabsComponents';
