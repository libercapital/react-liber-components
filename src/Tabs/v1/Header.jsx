import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

const Header = ({
  children, labels, slider, header, asList,
}) => (
  <Fragment>
    {
      asList
        ? children(labels, slider, header)
        : children(labels)
    }
  </Fragment>
);

Header.propTypes = {
  children: PropTypes.func,
  labels: PropTypes.oneOfType([PropTypes.any]),
  slider: PropTypes.oneOfType([PropTypes.any]),
  header: PropTypes.oneOfType([PropTypes.any]),
  asList: PropTypes.bool,
};

Header.defaultProps = {
  children: () => {},
  labels: null,
  slider: null,
  header: null,
  asList: false,
};

export default Header;
