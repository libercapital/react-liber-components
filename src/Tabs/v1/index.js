// eslint-disable-next-line import/no-cycle
export { default as Tabs } from './Tabs';
export { default as Tab } from './Tab';
export { default as TabHeader } from './Header';
