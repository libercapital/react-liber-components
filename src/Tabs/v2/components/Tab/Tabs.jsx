import React from 'react';
import CenterTabs from './CenterTabs';
import LeftTabs from './LeftTabs';

const Tabs = ({ type = 'center', ...props }) =>
  ({
    center: <CenterTabs {...props} />,
    left: <LeftTabs {...props} />,
  }[type]);

export default Tabs;
