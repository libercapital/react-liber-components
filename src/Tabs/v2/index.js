export { SimpleTab, SimpleTabs } from './components/SimpleTab';
export { Tabs } from './components/Tab';
export { default as DropDownTab } from './components/DropDownTab';
