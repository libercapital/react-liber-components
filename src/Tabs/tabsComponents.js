import { wrapper } from '../Util/VersionWrapper';

import * as v1 from './v1';
import * as v2 from './v2';

export const { TabHeader } = v1;

export const Tabs = wrapper([v1.Tabs, v2.Tabs]);

export const Tab = wrapper([v1.Tab, v2.SimpleTab]);

export const { SimpleTab, SimpleTabs, DropDownTab } = v2;

export { TabContainer, TabContent, TabContentDisplay } from './v2/components/Tab/Tabs.styles';
