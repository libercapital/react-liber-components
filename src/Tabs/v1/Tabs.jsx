import React, { Component } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import { Header, Label, Slide, BasicHeader } from './styles';
import { Tab } from '../tabsComponents';
import TabHeader from './Header';

function validateTab(props, propName, componentName) {
  let children = props[propName];
  if (children.constructor !== Array) {
    children = Array(children);
  }
  children.forEach(child => {
    if (child.type !== Tab && child.type !== TabHeader) {
      return new Error(`${componentName} only accept Tab or TabHeader as children.`);
    }
    return null;
  });
  return null;
}

class Tabs extends Component {
  constructor(props) {
    super(props);
    this.state = {
      current: props.active || 0,
      ...this.dispartChildren(props.children),
    };
  }

  static getDerivedStateFromProps(nextProps) {
    const { active } = nextProps;
    return active !== null && active !== undefined ? { current: nextProps.active } : {};
  }

  dispartChildren = children => {
    const tabChildren = React.Children.toArray(children).filter(child => child.type === Tab);
    const tabHeader = React.Children.toArray(children).filter(child => child.type === TabHeader)[0];
    return {
      tabChildren,
      tabHeader,
    };
  };

  handleSelection = index => {
    const { active } = this.props;
    if (active === null || active === undefined) {
      this.setState({ current: index });
    }
  };

  render() {
    const { current, tabChildren, tabHeader } = this.state;
    const { children, slideColor, ...props } = this.props;

    const labels = React.Children.toArray(tabChildren).map(child => child.props.label);
    const mappedLabels = labels.map((label, index) => {
      const id = _.uniqueId(label);
      return (
        <Label {...props} htmlFor={id} key={id} active={index === current} data-value={label}>
          <input
            onChange={() => this.handleSelection(index)}
            type="radio"
            name="tabs"
            id={id}
            checked={index === current}
          />
          {label}
        </Label>
      );
    });
    let header;
    if (tabHeader) {
      if (tabHeader.props.asList) {
        header = React.cloneElement(tabHeader, {
          labels: mappedLabels,
          slider: <Slide number={labels.length} current={current} color={slideColor} />,
          header: BasicHeader,
        });
      } else {
        header = React.cloneElement(tabHeader, {
          labels: (
            <BasicHeader>
              {mappedLabels}
              <Slide role="slider" number={labels.length} current={current} color={slideColor} />
            </BasicHeader>
          ),
        });
      }
    } else {
      header = (
        <Header>
          {mappedLabels}
          <Slide number={labels.length} current={current} color={slideColor} />
        </Header>
      );
    }

    return (
      <>
        {header}
        {tabChildren[current]}
      </>
    );
  }
}

Tabs.propTypes = {
  slideColor: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.string)]),
  children: validateTab,
  active: PropTypes.number,
};

Tabs.defaultProps = {
  slideColor: undefined,
  children: <div />,
  active: null,
};

export default Tabs;
