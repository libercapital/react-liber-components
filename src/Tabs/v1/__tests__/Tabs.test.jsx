import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import 'jest-styled-components';
import Tabs from '../Tabs';
import { Tab } from '../../tabsComponents';
import Header from '../Header';

describe('Testing Tabs Component', () => {
  let props;
  configure({ adapter: new Adapter() });

  beforeEach(() => {
    props = {
      label: 'test',
    };
  });
  it('Should render correctly', () => {
    const wrapper = shallow(
      <Tabs>
        <Tab {...props}>Test</Tab>
      </Tabs>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('Should render correctly with header', () => {
    const wrapper = shallow(
      <Tabs>
        <Header>{labels => labels}</Header>
        <Tab {...props}>Test</Tab>
      </Tabs>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('Should change state when change selection', () => {
    const wrapper = shallow(
      <Tabs>
        <Tab {...props}>Test</Tab>
        <Tab {...props}>Test2</Tab>
      </Tabs>,
    );
    const input = wrapper.find('input').at(1);
    expect(wrapper.state('current')).toBe(0);
    input.simulate('change');
    expect(wrapper.state('current')).toBe(1);
  });
});
