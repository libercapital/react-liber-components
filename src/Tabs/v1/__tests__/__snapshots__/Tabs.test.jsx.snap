// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Tabs Component Should render correctly 1`] = `
<Fragment>
  <styles__Header>
    <styles__Label
      active={true}
      data-value="test"
      htmlFor="test1"
      key="test1"
    >
      <input
        checked={true}
        id="test1"
        name="tabs"
        onChange={[Function]}
        type="radio"
      />
      test
    </styles__Label>
    <styles__Slide
      current={0}
      number={1}
    />
  </styles__Header>
  <ForwardRef
    key=".0"
    label="test"
  >
    Test
  </ForwardRef>
</Fragment>
`;

exports[`Testing Tabs Component Should render correctly with header 1`] = `
<Fragment>
  <Header
    asList={false}
    header={null}
    key=".0"
    labels={
      <ForwardRef(styles__BasicHeader)>
        <ForwardRef(styles__Label)
          active={true}
          data-value="test"
          htmlFor="test2"
        >
          <input
            checked={true}
            id="test2"
            name="tabs"
            onChange={[Function]}
            type="radio"
          />
          test
        </ForwardRef(styles__Label)>
        <ForwardRef(styles__Slide)
          current={0}
          number={1}
          role="slider"
        />
      </ForwardRef(styles__BasicHeader)>
    }
    slider={null}
  >
    <Component />
  </Header>
  <ForwardRef
    key=".1"
    label="test"
  >
    Test
  </ForwardRef>
</Fragment>
`;
